### PR TITLE
[Models CI] [Sweep Pipeline] Initial Sweep sampling tests

### PR DESCRIPTION
**Pull request by ifabijanic-tt**
Issue ticket: https://github.com/tenstorrent/tt-metal/issues/38316

Initial version of tt_sampling tests. Subject to further improvements and changes...

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:tt_sampling_tests)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:tt_sampling_tests)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:tt_sampling_tests)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:tt_sampling_tests)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:tt_sampling_tests)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:tt_sampling_tests)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:tt_sampling_tests)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:tt_sampling_tests)

### CODE REVIEW

**Review by sraizada-tt — COMMENTED** [review]
## Claude Review: v0 of TT Sampling Unit Tests

### Overall

Good first version. The test structure is clean — synthetic logits with controlled hot-token sets is the right approach for validating sampling behavior without model weights. The helper functions (`build_hot_logits`, `build_penalty_logits`, `run_sampling_generator`) make tests readable and composable. Coverage is broad: temperature, top-k, penalties (all 3 types), seeding, batch isolation, and mixed params.

### Issues

**1. Misleading class name: `TestPrefillWithDifferentParams`** (line 232)
These tests run decode-style sampling (single token from logits), not prefill. Prefill refers to processing the full prompt. Suggest renaming to `TestSamplingWithDifferentParams` or `TestDecodeWithDifferentParams`.

**2. `_broadcast` is dead code** (line 58-62)
Defined but never called anywhere in the file. Remove or use it.

**3. `top_p=0.0` in `test_prefill_topk_1_is_greedy`** (line 310)
```python
greedy_tokens = run_ttsampling_once(..., top_k=1, top_p=0.0, temperature=1.0)
```
`top_p=0.0` means "keep no tokens" in standard nucleus sampling semantics. If TTSampling doesn't special-case this, it could produce undefined behavior. Should be `top_p=1.0` (disabled) to isolate the top_k=1 greedy test.

**4. No dedicated temp=0 greedy correctness test**
There's no test asserting that `temperature=0` always picks the highest-logit token. `test_prefill_topk_1_is_greedy` tests `top_k=1` but not `temp=0`. Suggest adding:
```python
def test_greedy_picks_max_logit(self, mesh_device):
    args = make_sampling_args(mesh_device)
    logits = build_hot_logits(args, hot_tokens=[42, 43, 44])  # 42 has highest logit
    tokens = run_ttsampling_once(mesh_device, args, logits, top_k=32, top_p=1.0, temperature=0.0)
    assert all(tok == 42 for tok in tokens), "Greedy should always pick the max logit token"
```

**5. No dedicated top-p filtering test**
Top-p is only tested indirectly in mixed-param batches. There should be a test showing that `top_p < 1.0` actually restricts the output token set (e.g., with carefully crafted logits where top_p=0.5 should exclude certain tokens).

**6. Penalty tests are single-step only**
All penalty tests (`TestRepetitionPenaltyPerRequest`, `TestPresencePenaltyPerRequest`, `TestFrequencyPenaltyPerRequest`) run for `num_steps=1`. Missing coverage for **multi-step penalty accumulation** — e.g., does seeing a token 5 times penalize more than once (frequency), and does the penalty persist across decode steps?

**7. No multi-device tests for penalties/isolation**
Penalty tests and `TestBatchIsolation` are parametrized only with `mesh_device=[1]`. The batch isolation test is arguably the most important one to run on `(4,8)` — that's where cross-device contamination bugs would surface.

**8. Accessing private member `_seed_mapper`** (line 212)
```python
mesh_mapper=sg.seed_manager._seed_mapper,
```
This reaches into a private implementation detail. If `SeedManager` changes internally, this test breaks. Consider exposing a public property or method.

**9. `run_ttsampling_once` hardcodes `BATCH_SIZE`** (line 176)
```python
return extract_tokens(tt_tokens, BATCH_SIZE)
```
Should propagate the batch_size from the logits shape, or accept it as a parameter. If someone creates logits with `batch_size=2` via `build_hot_logits(..., batch_size=2)`, this will still read 32 tokens.

**10. Missing conftest.py / fixture documentation**
The `mesh_device` and `device_params` fixtures use `indirect=True` but aren't defined in this file. A comment noting where these fixtures come from (or a conftest.py included in the PR) would help.

**11. `device_params` parametrization inconsistency**
Only some tests specify `device_params` with `fabric_config`. For tests that don't specify it, what's the default? If fabric is required for multi-device, then `test_prefill_temperature_varied_in_batch` with `mesh_device=(4,8)` might silently use wrong fabric config.

**12. No tensor cleanup**
Each call to `run_sampling_generator` and `run_ttsampling_once` creates new `TTSampling`/`SamplingGenerator` objects with device tensors, but there's no explicit cleanup. Over ~30 test cases this could accumulate significant device memory. Consider adding a teardown or `del` + `ttnn.synchronize_device()`.

### Minor Nits

- Line 337: `penalties[i] <= 1.2` — the boundary is fragile. `1.2` is in the penalty list; this test assumes `rep_penalty=1.2` still picks the target token. This is implementation-dependent (depends on exact logit gap between target and base in `build_penalty_logits`).
- Each test class uses different magic token IDs (100s, 500s, 700s, 1000s, etc.) presumably to avoid collisions, but a comment explaining this convention would help.
- Line 265: `temperature = [0.0] * half + [1.5] * (BATCH_SIZE - half)` — `temperature=0.0` is greedy, then asserting `len(set(greedy)) == 1` tests determinism of same-input greedy, which is good, but could be made explicit in the test name.

### What's Good

- Synthetic logits approach is correct — no model weights needed
- Hot-token sets with descending logit values (`top_logit - rank * step`) give controlled stochastic distributions
- Seed reproducibility is tested thoroughly (8 seed-related tests)
- `per_user_hot_tokens` in `build_hot_logits` enables proper batch isolation testing
- `state_setup` callback pattern for penalty state is clean and composable
- Negative seed edge case test is a nice touch

### Summary

Solid v0. The main gaps are: (1) no greedy-correctness test, (2) no dedicated top-p test, (3) penalty tests need multi-step coverage, and (4) batch isolation should also run on multi-device. The naming (`Prefill`) should be fixed before it confuses future readers.

File: models/common/sampling/generator.py [hunk 1]
```diff
@@ -519,6 +525,22 @@ def reset_seed(self, seeds, user_ids):
             self.rngs[user].seed(seeds[i])
             self.seeds[user] = seeds[i]
 
+    def write_device_seed_values(self, seed_values):
+        if len(seed_values) != self.max_batch_size:
+            raise ValueError(f"Expected {self.max_batch_size} seed values, got {len(seed_values)}")
+        try:
+            wrapped = [int(seed) & 0xFFFFFFFF for seed in seed_values]
```
Comment (djordje-tt) [inline models/common/sampling/generator.py:786]: why do we have 0xFFFFFFFF?

Comment (ifabijanic-tt) [reply models/common/sampling/generator.py:786]: Good question and this will probably need a comment in the code since it's not that obvious. Since device seed buffer is defined as uint32 in tt_sampling.py file this bitwise AND ensures that all seed numbers are in the uint32 range because it just takes the lowest 32 bits from any integer number. This makes integer number conversion predictable, explicit and localized here. It's also equivalent as taking modulo with "% 2**32" but this should (theoretically) be a little bit faster. If needed it can also we rewritten as "int(seed) % 2 **32) for readability purposes...

**Comment by mtairum** [comment]
This is currently delayed due to other P0 issues. We'll revisit shortly.

**Comment by tchedaTT** [comment]
Please address copilot's comments

**Comment by ifabijanic-tt** [comment]
This PR will be closed as @fbuticTT will continue working on this issue https://github.com/tenstorrent/tt-metal/issues/38316. The same branch will be used but new PR with better description, potential fixes, new rebase and Filips ownership will be created.
